os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -189,91 +176,53 @@ int binary_manager_load_binary(int bin_idx)
 		}
 	}
 
-	/* Failed to find valid binary */
-	if (loadable == false) {
-		bmdbg("Failed to find loadable binary %d\n", bin_idx);
+	if (valid_bin_count == 0) {
+		bmdbg("Failed to find valid header of binary %s\n", BIN_NAME(bin_idx));
 		return ERROR;
 	}
 
 	/* Load binary */
 	do {
-		if (is_new_bin == false) {
-			strncpy(load_attr.bin_name, header_data[latest_idx].bin_name, BIN_NAME_MAX);
-			snprintf(devname, BINMGR_DEVNAME_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, latest_idx));
-			load_attr.bin_size = header_data[latest_idx].bin_size;
-			load_attr.compression_type = header_data[latest_idx].compression_type;
-			load_attr.ram_size = header_data[latest_idx].bin_ramsize;
-			load_attr.stack_size = header_data[latest_idx].bin_stacksize;
-			load_attr.priority = header_data[latest_idx].bin_priority;
-			load_attr.offset = CHECKSUM_SIZE + header_data[latest_idx].header_size;		
-			is_new_bin = true;
-		}
-		bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, load_attr.bin_size, load_attr.offset);
+		strncpy(load_attr.bin_name, header_data[latest_idx].bin_name, BIN_NAME_MAX);
+		snprintf(devname, BINMGR_DEVNAME_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, latest_idx));
+		load_attr.bin_size = header_data[latest_idx].bin_size;
+		load_attr.compression_type = header_data[latest_idx].compression_type;
+		load_attr.ram_size = header_data[latest_idx].bin_ramsize;
+		load_attr.stack_size = header_data[latest_idx].bin_stacksize;
+		load_attr.priority = header_data[latest_idx].bin_priority;
+		load_attr.offset = CHECKSUM_SIZE + header_data[latest_idx].header_size;
 
-		/* If a priority of loaded task is higher than priority of loading thread,
-		 * we need to lock scheduling until updating loading info in binary table. */
-		if (load_attr.priority >= LOADINGTHD_PRIORITY) {
-			sched_lock();
-			is_sched_locked = true;
-		}
+		bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, load_attr.bin_size, load_attr.offset);
 
-		ret = load_binary(devname, &load_attr);
-		if (ret > 0) {
-			bin_pid = (pid_t)ret;
-			bmvdbg("Load '%s' success! pid = %d\n", devname, bin_pid);
-			break;
-		} 
-#if (defined(CONFIG_BUILD_PROTECTED) || defined(CONFIG_BUILD_KERNEL)) && defined(CONFIG_MM_KERNEL_HEAP)
-		else if (errno == ENOMEM) {
-			if (is_sched_locked) {
-				sched_unlock();
-				is_sched_locked = false;
+		retry_count = 0;
+		
+		while (retry_count < BINMGR_LOADING_TRYCNT) {
+			ret = load_binary(bin_idx, devname, &load_attr);
+			if (ret > 0) {
+				bmvdbg("Load '%s' success! pid = %d\n", devname, ret);
+				/* Set the data in table from header */
+				BIN_USEIDX(bin_idx) = latest_idx;
```

Comment:
Please update these information before task activation.

Reply:
No need to be updated ahead because this index is for partition. We read all partitions and load the latest binary regardless of in-use index when fault.
I think binary index and pid of main task only should be updated before task activation.